### PR TITLE
Optimize matmuls involving block diagonal matrices

### LINKED ISSUE #1044: Add rewrite to optimize `block_diag(a, b) @ c`
### Description

Given a block diagonal matrix formed of matrices $A \in \mathbb R^{n \times m}$ and $B^{o \times p}$, such that:

$$ D = \begin{bmatrix} 
A & 0 \\ 
0 & B 
\end{bmatrix} $$

Computing the matrix multiplication $DC$ can be simplified. Define $C_1 \in \mathbb R^{n \times k}$ and $C_2^{o \times k}$ such that: 

$$
C = \begin{bmatrix} C_1 \\
C_2 \end{bmatrix}
$$

then:

$$
\begin{align}
DC &= 
\begin{bmatrix} 
A & 0 \\
0 & B 
\end{bmatrix} 
\begin{bmatrix} 
C_1 \\ 
C_2 
\end{bmatrix} \\
&= \begin{bmatrix}
A C_1 \\
 B C_2 \end{bmatrix}
\end{align}
$$

We can compute these smaller dot products then concatenate the results back together for a speedup. Code:

```python
import numpy as np
from scipy import linalg

rng = np.random.default_rng()
n = 1000
A, B = rng.normal(size=(2, n, n))
C = rng.normal(size=(2*n, 2*n))

A, B, C = map(np.ascontiguousarray, [A, B, C])

def direct(A, B, C):
    X = linalg.block_diag(A, B)
    return X @ C

def rewrite(A, B, C):
    n = A.shape[0]
    C_1, C_2 = C[:n], C[n:]
    return np.concatenate([A @ C_1, B @ C_2])
    

np.allclose(direct(A, B, C), rewrite(A, B, C)) # True
```

Speed test:

```python
direct_time = %timeit -o direct(A, B, C)
rewrite_time = %timeit -o rewrite(A, B, C)

speedup_factor = (1 - rewrite_time.best / direct_time.best)
print(f'{speedup_factor:0.2%}')

75.8 ms ± 3.05 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
43.5 ms ± 1.62 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
40.77%
```

### PR DESCRIPTION
**Pull request by jessegrabowski**
## Description
This PR adds a rewrite to optimize matrix multiplication involving block diagonal matrices. When we have a a matrix `X = BlockDiag(A, B)`, when you do `Z = X @ Y`, there's no interaction between terms in the `A` part and `B` part of the `X` matrix. So the dot can be instead computed as `row_stack(A @ Y[:X.shape[0]], B @ Y[X.shape[0]:]` (or in the general case, `Y` can be split into `n` pieces with appropriate shapes, and do `row_stack([diag_component @ y_split for diag_component, y_split in zip(BlockDiag.inputs, split(Y, *args)])`. If the case where the blockdiag matrix is right-multiplying, you instead col_stack and slice on axis=1. 

Anyway, it's a lot faster to do this, because matmuls scale really badly in the dimension of the input, so doing two smaller operations is preferred. Here are the benchmarks, small has `n=10`, medium has `n=100`, large has `n=1000`. But in all cases it shows at least 2x speedup.

```
---------------------------------------------------------------------------------------------------------------- benchmark: 6 tests ----------------------------------------------------------------------------------------------------------------
Name (time in us)                                                       Min                   Max                  Mean              StdDev              Median                 IQR             Outliers           OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_block_diag_dot_to_dot_concat_benchmark[small-rewrite]           4.5830 (1.0)         90.7090 (2.13)         5.3007 (1.0)        1.6155 (2.53)       5.2080 (1.0)        0.1660 (1.0)         67;560  188,654.6546 (1.0)       12533           1
test_block_diag_dot_to_dot_concat_benchmark[small-no_rewrite]        8.5420 (1.86)        90.1670 (2.12)        10.1183 (1.91)       1.6055 (2.51)      10.0000 (1.92)       0.1680 (1.01)      430;2150   98,830.6599 (0.52)      18721           1

test_block_diag_dot_to_dot_concat_benchmark[medium-rewrite]          6.1250 (1.34)        44.8750 (1.05)         7.2724 (1.37)       0.6386 (1.0)        7.4170 (1.42)       0.2490 (1.50)     7575;7886  137,505.3510 (0.73)      35875           1
test_block_diag_dot_to_dot_concat_benchmark[medium-no_rewrite]      14.0420 (3.06)        42.6250 (1.0)         16.5707 (3.13)       1.3341 (2.09)      17.2500 (3.31)       2.1660 (13.05)     1174;108   60,347.4538 (0.32)      12177           1

test_block_diag_dot_to_dot_concat_benchmark[large-rewrite]          14.6660 (3.20)       248.2920 (5.83)        16.5375 (3.12)       4.7284 (7.40)      16.1250 (3.10)       0.4590 (2.76)      249;1621   60,468.5555 (0.32)      18765           1
test_block_diag_dot_to_dot_concat_benchmark[large-no_rewrite]      788.6250 (172.08)   1,982.7500 (46.52)    1,019.2728 (192.29)   150.6524 (235.91)   987.3335 (189.58)   130.6250 (786.86)      132;63      981.0916 (0.01)        734           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

## Related Issue

- [ ] Closes #1044 #1552
- [ ] Related to 

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pytensor--1493.org.readthedocs.build/en/1493/

### CODE REVIEW

File: pytensor/tensor/rewriting/math.py [hunk 1]
```diff
@@ -167,6 +170,65 @@ def local_0_dot_x(fgraph, node):
...
+    Perform the rewrite ``dot(block_diag(A, B), C) -> concat(dot(A, C), dot(B, C))``
+
+    BlockDiag results in the creation of a matrix of shape ``(n1 * n2, m1 * m2)``. Because dot has complexity
+    of approximately O(n^3), it's always better to perform two dot products on the smaller matrices, rather than
+    a single dot on the larger matrix.
+    """
+    x, y = node.inputs
+    op = node.op
+
+    def check_for_block_diag(x):
```
Comment (ricardoV94) [inline pytensor/tensor/rewriting/math.py]: Since this is likely very rare compared to Dot, may it be worth to track the rewrite on BlockDiagonal instead? Like we did in https://github.com/pymc-devs/pytensor/pull/1483/files with `expm1` and the use `clients_at_depth` and use a return dictionary

File: pytensor/tensor/rewriting/math.py [hunk 2]
```diff
@@ -167,6 +170,65 @@ def local_0_dot_x(fgraph, node):
             return [constant_zero]
 
 
+@register_canonicalize
+@register_specialize
+@register_stabilize
+@node_rewriter([Dot])
```
Comment (ricardoV94) [inline pytensor/tensor/rewriting/math.py]: Why Dot and not BlockwiseDot? Also I wouldn't consider this a specialization

File: pytensor/tensor/rewriting/math.py [hunk 3]
```diff
@@ -167,6 +170,65 @@ def local_0_dot_x(fgraph, node):
...
+            axis=1,
+        )
+
+        new_components = [
+            op(x_split, component) for component, x_split in zip(components, x_splits)
+        ]
+        new_output = join(1, *new_components)
+
+    else:
+        return None
```
Comment (ricardoV94) [inline pytensor/tensor/rewriting/math.py]:
```suggestion
        # Both inputs are block_diagonal. Do nothing
        return None
```

**Review by ricardoV94 — COMMENTED** [review]
Looks great! Some minor optimization questions

File: pytensor/tensor/rewriting/math.py (hunk 1 above)
Comment (jessegrabowski) [reply pytensor/tensor/rewriting/math.py]: Yes, good idea. I didn't have a good sense of the right thing to track. This is also why it doesn't track BlockwiseDot

File: pytensor/tensor/rewriting/math.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/tensor/rewriting/math.py]: Well maybe it's better as specialization than canonicalization (shouldn't be both is the point), but we have the awkward thing that the blas operations come before specialization...

We should probably try to put them after but that's not in scope for sure.

Comment (jessegrabowski) [reply pytensor/tensor/rewriting/math.py]: I'll go with stabilize for now then? BlasOpt is at 1.7 and stabilize is at 1.5.

I agree the best label in general is specialize. But it's also a canonicalization of sorts, since this should be the preferred representation for this graph.

File: pytensor/tensor/rewriting/math.py [hunk 4]
```diff
@@ -167,6 +170,76 @@ def local_0_dot_x(fgraph, node):
...
+
+    def check_for_block_diag(x):
+        return x.owner and (
+            isinstance(x.owner.op, BlockDiagonal)
+            or isinstance(x.owner.op, Blockwise)
+            and isinstance(x.owner.op.core_op, BlockDiagonal)
+        )
+
+    # Check that the BlockDiagonal is an input to a Dot node:
+    clients = list(get_clients_at_depth(fgraph, node, depth=1))
```
Comment (ricardoV94) [inline pytensor/tensor/rewriting/math.py]: You should iterate over the clients and return when there's a match (just indent the code below inside the loop). Why is it a problem if the BlockDiagonal has more clients?

Comment (jessegrabowski) [reply pytensor/tensor/rewriting/math.py]: Why would it though?

File: pytensor/tensor/rewriting/math.py [hunk 5]
```diff
@@ -167,6 +170,77 @@ def local_0_dot_x(fgraph, node):
...
+        return x.owner and (
+            isinstance(x.owner.op, BlockDiagonal)
+            or isinstance(x.owner.op, Blockwise)
+            and isinstance(x.owner.op.core_op, BlockDiagonal)
+        )
+
+    # Check that the BlockDiagonal is an input to a Dot node:
+    for client in get_clients_at_depth(fgraph, node, depth=1):
+        if not isinstance(client.op, Dot):
+            return
```
Comment (ricardoV94) [inline pytensor/tensor/rewriting/math.py]:
```suggestion
            continue
```

Comment (ricardoV94) [reply pytensor/tensor/rewriting/math.py]: all the inner returns become continue, except the replacement one

File: pytensor/tensor/rewriting/math.py [hunk 6]
```diff
@@ -167,6 +170,77 @@ def local_0_dot_x(fgraph, node):
...
+        x, y = client.inputs
+
+        if not (check_for_block_diag(x) or check_for_block_diag(y)):
+            continue
+
+        # Case 1: Only one input is BlockDiagonal. In this case, multiply all components of the block-diagonal with the
+        # non-block diagonal, and return a new block diagonal
+        if check_for_block_diag(x) and not check_for_block_diag(y):
+            components = x.owner.inputs
+            y_splits = split(
```
Comment (ricardoV94) [inline pytensor/tensor/rewriting/math.py]: Isn't this and the join along the 0th axis assuming a BlockwiseBlockDiagonal without batch dims?

Also not sure why you look for Dot but not Blockwise of _matrix_matrix_matmul. It doesn't always get rewritten as a Dot.

Comment (ricardoV94) [reply pytensor/tensor/rewriting/math.py]: I guess because you don't look for batch dot you can only have a BlockDiagonal without batch dims. That's fine but maybe a bit implicit. You can also wait for the useless BlockwiseBlockdiagonal to be rewritten as BlockDiagonal and only track that.

More importantly because you track a regular dot and not the matmul you may have a vector \* matrix or matrix \* vector product. Does the rewrite handle these correctly?

Comment (jessegrabowski) [reply pytensor/tensor/rewriting/math.py]: If this is what's implied it's not on purpose. I'll modify it to account for blockwise dot.

There's no canonical `dot` form we rewrite to in an intermediate step to make reasoning about graphs easier? It seems nuts to have to to look for a bunch of different `_matrix_matrix_matmul` or `_matrix_vec_matmul` or `dot22` or whatever.

Comment (ricardoV94) [reply pytensor/tensor/rewriting/math.py]: I'm simplifying every blockwise as blockwise 2x2 dot (ie matmul) in https://github.com/pymc-devs/pytensor/pull/1471

The dot22 and dot22scalar are stuff from the blas pipeline and I've been hesitant to touch it. As first steps I would like to move them after specialize and to get rid of dot22scalar (should just be gemm). Those blas stuff should also work with blockwise but they currently don't.

Anyway if you target blockwise and core 2x2dot in this PR that should be the most robust going forward even if it misses some cases now. I suggest you explicitly exclude the vector matrix dots for now.

Comment (jessegrabowski) [reply pytensor/tensor/rewriting/math.py]: skipping matrix-vector is a bit of a bummer

Comment (ricardoV94) [reply pytensor/tensor/rewriting/math.py]: then make sure the rewrite works or wait for the PR

Comment (jessegrabowski) [reply pytensor/tensor/rewriting/math.py]: you're like the teacher from that movie about jazz drumming u-u

Comment (ricardoV94) [reply pytensor/tensor/rewriting/math.py]: which movie? whiplash?

**Comment by jessegrabowski** [comment]
I'm still struggling with ExpandDims "blocking" between Dot and BlockDiag in the batch case here, do you have a suggestion for how to handle it?

**Comment by ricardoV94** [comment]
What sort of graph exactly?

**Comment by jessegrabowski** [comment]
Actually the issue is the Join in the blockwise case. Given a graph like:

```
Matmul [id A] <Tensor3(float64, shape=(3, 10, 10))> 4
 ├─ ExpandDims{axis=0} [id B] <Tensor3(float64, shape=(1, 10, 10))> 3
 │  └─ d [id C] <Matrix(float64, shape=(10, 10))>
 └─ Blockwise{BlockDiagonal{n_inputs=3}, (m0,n0),(m1,n1),(m2,n2)->(m,n)} [id D] <Tensor3(float64, shape=(3, 10, 10))> 2
    ├─ ExpandDims{axis=0} [id E] <Tensor3(float64, shape=(1, 4, 2))> 1
    │  └─ a [id F] <Matrix(float64, shape=(4, 2))>
    ├─ b [id G] <Tensor3(float64, shape=(3, 2, 4))>
    └─ ExpandDims{axis=0} [id H] <Tensor3(float64, shape=(1, 4, 4))> 0
       └─ c [id I] <Matrix(float64, shape=(4, 4))>
```

We pull out the 3 inputs and split `d` into [(1, 10, 4), (1, 10, 2), (1, 10, 4)], then do the 3 dots, obtaining results of shapes `[(1, 10, 2), (3, 10, 4), (1, 10, 4)]`

This should all be correct, but now the join fails, because the batch dim on the 3 results is not the same. Looking at it in the debugger, it seems like i need to tile `a` and `c` along the broadcastable dims until their shapes match up. Is there a helper for that? 

To clarify that last point, here is `res[:, 0, :2]` from the matmul, which is the contribution from the `(1, 10, 2)` component:

```
array([[ 0.30917466, -1.10133142],
       [ 0.30917466, -1.10133142],
       [ 0.30917466, -1.10133142]])
```

And the `(1, 10, 2)` component itself:

```
array([[[ 0.30917466, -1.10133142],
        [-2.7780056 ,  6.51444629],
        [ 0.79439975, -2.1491907 ],
        [-3.09616644, -0.1655169 ],
        [ 0.28896778,  0.65660826],
        [ 2.88199648, -3.4479807 ],
        [-2.07519668,  1.33874492],
        [-0.57444061, -3.40710469],
        [-2.41102468,  2.96805356],
        [ 1.06383867, -0.90597861]]])
```

As you can see, it's just the first row duplicated 3 times.

**Comment by ricardoV94** [comment]
> Is there a helper for that?

That's what I'm proposing in the first part of #1552 ;)

**Comment by jessegrabowski** [comment]
I added a `concat_with_broadcast` helper, basically just copying your code from #1552

The last case I'm still struggling with is when there are Dimshuffles between the Dot and the BlockDiag, for example:

```
Transpose{axes=[2, 0, 1, 3]} [id A] 10
 └─ Reshape{4} [id B] 9
    ├─ Dot [id C] 8
    │  ├─ Squeeze{axis=0} [id D] 7
    │  │  └─ Reshape{3} [id E] 6
    │  │     ├─ Blockwise{BlockDiagonal{n_inputs=3}, (m0,n0),(m1,n1),(m2,n2)->(m,n)} [id F] 5
    │  │     │  ├─ ExpandDims{axis=0} [id G] 4
    │  │     │  │  └─ a [id H]
    │  │     │  ├─ b [id I]
    │  │     │  └─ ExpandDims{axis=0} [id J] 3
    │  │     │     └─ c [id K]
    │  │     └─ [ 1 -1 10] [id L]
    │  └─ Squeeze{axis=0} [id M] 2
    │     └─ Reshape{3} [id N] 1
    │        ├─ Transpose{axes=[1, 2, 0, 3]} [id O] 0
    │        │  └─ d [id P]
    │        └─ [ 1 10 -1] [id Q]
    └─ [ 3 10  3 10] [id R]
```

I put the rewrite into specialize because I was hoping by that point these dimshuffles would have been cleaned up, but no such luck.

**Comment by jessegrabowski** [comment]
I cobbled together something that works, but not sure if it's ideal

**Comment by ricardoV94** [comment]
Yeah you want the rewrite before the batched_to_core_dot in specialize. Register it in canonicalize or stabilize?

**Comment by jessegrabowski** [comment]
It's in stabilize now

File: pytensor/tensor/extra_ops.py [hunk 7]
```diff
@@ -2018,6 +2018,42 @@ def broadcast_with_others(a, others):
     return brodacasted_vars
 
 
+def concat_with_broadcast(tensor_list, dim=0):
```
Comment (ricardoV94) [inline pytensor/tensor/extra_ops.py]:
```suggestion
def concat_with_broadcast(tensor_list, axis=0):
```

File: pytensor/tensor/extra_ops.py [hunk 8]
```diff
@@ -2018,6 +2018,42 @@ def broadcast_with_others(a, others):
...
+    """
+    Concatenate a list of tensors, broadcasting the non-concatenated dimensions to align.
+    """
+    dim = dim if dim > 0 else tensor_list[0].ndim + dim
```
Comment (ricardoV94) [inline pytensor/tensor/extra_ops.py]: You should assert everything has the same ndim. Then use the numpy normalize_axis_index to handle the negative indices + useful error message if it's invalid

File: pytensor/tensor/extra_ops.py [hunk 9]
```diff
@@ -2018,6 +2018,42 @@ def broadcast_with_others(a, others):
...
+    non_concat_shape = [None] * tensor_list[0].ndim
+
+    # If all inputs have the same broadcastable dim, it's not really a broadcast
+    all_bcast = [
```
Comment (ricardoV94) [inline pytensor/tensor/extra_ops.py]: Set broadcast shape to all 1 (except axis), then you don't need the all_bcast logic? If you never overwrite the shape it was because everything was bcast on that axis

File: pytensor/tensor/extra_ops.py (hunk 7 above)
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py]: You should be able to now reuse this helper on the xtensor rewrite?

File: pytensor/tensor/rewriting/math.py [hunk 10]
```diff
@@ -146,6 +149,72 @@ def local_0_dot_x(fgraph, node):
...
+        return
+
+    # Check that the BlockDiagonal is an input to a Dot node:
+    for client in itertools.chain.from_iterable(
+        get_clients_at_depth(fgraph, node, depth=i) for i in [1, 2]
+    ):
+        if not (
+            (
+                isinstance(client.op, Dot)
+                and all(input.ndim == 2 for input in client.inputs)
```
Comment (ricardoV94) [inline pytensor/tensor/rewriting/math.py]: This check is no longer needed, dot is always 2d now.

You can just do `client.op in (_dot, _matmul)`

File: tests/tensor/test_extra_ops.py [hunk 11]
```diff
@@ -1333,3 +1333,33 @@ def test_space_ops(op, dtype, start, stop, num_samples, endpoint, axis):
...
+    c_val = fn(a_val, b_val)
+
+    # The result should be a tile + concat
+    np.testing.assert_allclose(c_val[:, :, :5], np.tile(a_val, (5, 1, 1)))
+    np.testing.assert_allclose(c_val[:, :, 5:], b_val)
+
+    # If a and b already conform, the result should be the same as a concatenation
+    a = pt.tensor("a", shape=(1, 1, 3, 5, 10))
+    b = pt.tensor("b", shape=(1, 1, 3, 2, 10))
+    c = pt.concatenate([a, b], axis=-2)
```
Comment (ricardoV94) [inline tests/tensor/test_extra_ops.py]: You still want to test the new function. You should also test it does the right thing with one or zero inputs

Comment (jessegrabowski) [reply tests/tensor/test_extra_ops.py]: oh that was an autocomplete flub. how embarrassing.

**Review by ricardoV94 — APPROVED** [review]
Looks great 

I suspect we'll be missing some cases when there's a Dimshuffle between the BlockDiag and the Dot like

`block_diag(...) @ tensor3` where the block diag will have an expand_dims that I don't think is lifted into the `block_diag`?

Also we should rewrite `block_diag(...).mT -> block_diag(e0.mT, ..., eN.mT)` to lift the transpose through the inputs. As well as some `elemwise(block_diag(...))` like `mul`.

In general we should have a list of elemwise functions that map zero->zero, this comes up in a lot of these linalg rewrites.

All this should be a separate issue, this PR is concise and clean like this.

**Comment by ricardoV94** [comment]
Is this related to the COLA issue as well (to link if so)?

File: pytensor/tensor/extra_ops.py [hunk 12]
```diff
@@ -2018,6 +2018,42 @@ def broadcast_with_others(a, others):
...
+    for tensor_inp in tensor_list:
+        for i, (bcast, sh) in enumerate(
+            zip(tensor_inp.type.broadcastable, tensor_inp.shape)
+        ):
+            if bcast or i == axis:
+                continue
+            non_concat_shape[i] = sh
+
+    assert non_concat_shape.count(None) == 1
+
```
[the remark this answers is not in the document]
Comment (jessegrabowski) [reply pytensor/tensor/extra_ops.py:2047]: This is sanity check and should never be triggered. An assert is appropriate in this case.

**Comment by jessegrabowski** [comment]
> Is this related to the COLA issue as well (to link if so)?

We don't have this one on the COLA list, because we didn't have a subsection for `dot`. I will add it.
